llvm/include/llvm/CodeGen/BasicTTIImpl.h

```diff
@@ -285,6 +286,64 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
     return false;
   }
 
+  /// Several intrinsics struct-ret (including llvm.sincos[pi] and llvm.modf)
+  /// can be lowered to a vector library call (for certain VFs). The vector
+  /// library functions correspond to the scalar calls (e.g. sincos or modf),
+  /// which unlike the intrinsic return values via output pointers. This helper
+  /// checks if a vector call exists for the given intrinsic, and returns the
+  /// cost, which includes the cost of the mask (if required), and the loads for
+  /// values returned via output pointers. \p LC is the scalar libcall and
+  /// \p CallRetElementIndex (optional) is the struct element which is mapped to
+  /// the call return value. If std::nullopt is returned, the no vector library
+  /// call is available, so the intrinsic should be assigned the default cost
+  /// (e.g. scalarization).
+  std::optional<InstructionCost> getMultipleResultIntrinsicVectorLibCallCost(
+      const IntrinsicCostAttributes &ICA, TTI::TargetCostKind CostKind,
+      RTLIB::Libcall LC, std::optional<unsigned> CallRetElementIndex = {}) {
+    Type *RetTy = ICA.getReturnType();
+    // Vector variants of the intrinsic can be mapped to a vector library call.
+    auto const *LibInfo = ICA.getLibInfo();
+    if (!LibInfo || !isa<StructType>(RetTy) ||
+        !isVectorizedStructTy(cast<StructType>(RetTy)))
+      return std::nullopt;
+
+    // Find associated libcall.
+    const char *LCName = getTLI()->getLibcallName(LC);
+    if (!LC || !LCName)
```

Comment:
Hmm, I'm not sure I understand why you check for `!LC` here because LC=0 seems to be a valid enum. See llvm/include/llvm/IR/RuntimeLibcalls.h:

```
enum Libcall {
#define HANDLE_LIBCALL(code, name) code,
#include "llvm/IR/RuntimeLibcalls.def"
#undef HANDLE_LIBCALL
};
```

where the first entry in `llvm/include/llvm/IR/RuntimeLibcalls.def` is:

```
// Integer
HANDLE_LIBCALL(SHL_I16, "__ashlhi3")
```

Reply:
Good spot! Removed the erroneous !LC check :+1: